CRM/Core/DAO.php

```diff
@@ -935,6 +935,13 @@ public static function writeRecord(array $record): CRM_Core_DAO {
     }
     $entityName = CRM_Core_DAO_AllCoreTables::getBriefName($className);
 
+    // For legacy reasons, empty values would sometimes be passed around as the string 'null'.
+    // The DAO treats 'null' the same as '', and an empty string makes a lot more sense!
```

Comment:
I have a memory that it might depend on data type and maybe more so whether the field is "NOT NULL". If the field doesn't have NOT NULL, then I'm thinking there is a difference and 'null' sets it to null, and '' sets it to ''.

And the reason it uses string 'null' is because real null means "don't change the field's value".

I haven't confirmed the above, just that's what I remember.

Reply:
For reference, since  236737d5c7cff529f06948a6994d284b81f4d5b7 APIv4 always converts `NULL` to `''` before saving values to the BAO, and we haven't had any problems with it. This just moves it one layer down so that hooks see consistent values.

Reply:
Ok I see. Fiddling around a bit it seems you end up with the same amount of crazy either way, so yes it's better for hooks.

What's amusing is it converts `''` back to `'null'` later in some situations which then in a smaller subset of those situations stores the literal string 'null' when it really should be null, but that's for another day.